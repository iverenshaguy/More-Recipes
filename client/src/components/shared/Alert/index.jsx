import NormalAlert from './NormalAlert';

export default { NormalAlert };
