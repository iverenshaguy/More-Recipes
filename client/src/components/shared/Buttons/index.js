import AddRecipeBtn from './AddRecipeBtn';

export default { AddRecipeBtn };
