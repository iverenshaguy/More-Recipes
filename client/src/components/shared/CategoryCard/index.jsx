import React from 'react';
import { Link } from 'react-router-dom';
import { Card, CardImg, CardImgOverlay } from 'reactstrap';
import './CategoryCard.scss';

const CategoryCard = () => (
  <div className="col-xs-12 col-sm-6 col-md-4 col-lg-3 mt-4 category-card">
    <Link to="/">
      <Card className="m-0 p-0 card-image">
        <CardImg
          bottom
          width="100%"
          className="img-fluid"
          src="images/jollof-rice-img.jpg"
          alt="recipe"
        />
        <CardImgOverlay className="text-center card-image-overlay text-white px-2 recipe-rating">
          <h5 className="mb-1 pt-3">Category</h5>
        </CardImgOverlay>
      </Card>
    </Link>
  </div>
);

export default CategoryCard;
