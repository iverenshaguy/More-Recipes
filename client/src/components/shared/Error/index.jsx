import React from 'react';
import PropTypes from 'prop-types';
import './Error.scss';

/**
 * @exports
 * @function Error
 * @param {object} props
 * @returns {JSX} Error
 */
const Error = (props) => {
  const { type } = props;
  return (
    <div className="container align-items-center justify-content-center" id="error-wrapper">
      <div className="container">
        <h3 className="text-center">{type === 404 ? 'Page Not Found' : 'Something happened, please try again'}</h3>
      </div>
    </div>
  );
};

Error.propTypes = {
  type: PropTypes.number
};

Error.defaultProps = {
  type: 500
};

export default Error;
