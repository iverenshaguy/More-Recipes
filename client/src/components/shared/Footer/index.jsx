import React from 'react';
import './Footer.scss';

/**
 * @exports
 * @function Footer
 * @extends Component
 * @returns {JSX} Footer
 */
const Footer = () => (
  <footer className="footer">
    <div className="container">
      <p>Â© {(new Date()).getFullYear()} More Recipes</p>
    </div>
  </footer>
);

export default Footer;
