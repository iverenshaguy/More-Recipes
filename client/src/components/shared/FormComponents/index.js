import RenderInput from './RenderInput';
import RenderSelect from './RenderSelect';
import RenderFileInput from './RenderFileInput';
import RenderCheckBox from './RenderCheckBox';
import AddRecipeCompA from './AddRecipeCompA';
import AddRecipeCompB from './AddRecipeCompB';

export default {
  RenderInput,
  RenderSelect,
  RenderCheckBox,
  AddRecipeCompA,
  AddRecipeCompB,
  RenderFileInput
};
