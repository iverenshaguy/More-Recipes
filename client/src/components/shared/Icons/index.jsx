import RenderIcon from './RenderIcon';
import RenderVoteIcon from './RenderVoteIcon';

export default { RenderIcon, RenderVoteIcon };
