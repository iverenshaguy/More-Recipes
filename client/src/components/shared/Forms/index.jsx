import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form as BootstrapForm, Button } from 'reactstrap';
import { asyncValidate, syncValidate } from '../../../helpers/validations';
import AuthForm from './AuthForm';
import { NormalAlert } from '../Alert';
import { arrayToObject } from '../../../utils';
import { formHelpers } from '../../../helpers';

/**
 * @exports
 * @class Form
 * @extends Component
 * @returns {component} Form
 */
class Form extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      title: PropTypes.string,
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);
    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];

    this.state = {
      type,
      values: arrayToObject(fields, ''),
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.validateField = this.validateField.bind(this);
    this.validateForm = this.validateForm.bind(this);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors() {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }


  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      values: { ...this.state.values, [name]: value },
      touched: { ...this.state.touched, [name]: true },
      pristine: false,
    }, () => { this.validateField(name, value); });
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  handleFocus() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur(event) {
    const { target } = event;
    const { value, name } = target;
    const { type } = this.state;

    this.setState({
      touched: { ...this.state.touched, [name]: true },
    }, () => { this.validateField(name); });

    setTimeout(() => {
      if ((name === 'email' || name === 'username') && this.state.error[name] === null) {
        // only set asyncvalidating when sync validation is completed
        this.setState({ asyncValidating: true }, () => {
          asyncValidate(type)(name, value)
            .then(() => { this.setState({ asyncValidating: false }); })
            .catch((error) => {
              this.setState({
                error: { ...this.state.error, [name]: error[name] },
                asyncValidating: false
              });
            });
        });
      }
    }, 500);
  }

  /**
   * @memberof Form
   * @param {string} field
   * @param {string} values
   * @returns {nothing} Returns nothing
   */
  validateField(field) {
    const { type } = this.state;
    const error = syncValidate(type)(field, this.state.values);

    if (error) {
      this.setState({
        error: {
          ...this.state.error,
          [field]: error
        }
      }, this.validateForm);
    } else {
      this.setState({
        error: {
          ...this.state.error,
          [field]: null
        }
      }, this.validateForm);
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm() {
    const formErrorArrayLength =
      Object.values(this.state.error).filter(value => value !== null).length;
    const formTouchedArrayLength =
      Object.values(this.state.touched).filter(value => value === true).length;
    const formFieldCount = Object.keys(this.state.values).length;

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength && formTouchedArrayLength === formFieldCount) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;

    this.props.dispatch(formSubmitMapper[type](this.state.values));
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const {
      values, touched, error, pristine, formValid, asyncValidating
    } = this.state;

    const {
      submitting, submitError, type, meta
    } = this.props;

    const { title, btnText, extra } = meta;

    const formState = {
      values,
      touched,
      error,
      pristine,
      formValid,
      asyncValidating
    };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    return (
      <div>
        <h4 className="text-center">{title}</h4>
        <hr />
        {type !== 'login' &&
          <p className="text-muted mx-auto text-center">
            Fields marked
            <span className="text-danger">*</span> are important
          </p>}
        <BootstrapForm className="mt-4 mb-3 px-5" onSubmit={handlers.handleSubmit}>
          {submitError && (
            <NormalAlert color="danger">
              <p className="text-center mb-0">{submitError}</p>
            </NormalAlert>
          )}
          {(type === 'signup' || type === 'login') && <AuthForm type={type} state={formState} handlers={handlers} />}
          <Button className="btn-block mt-0" disabled={!formValid || pristine || submitting}>
            {btnText}
          </Button>
        </BootstrapForm>
        {extra && extra}
      </div>);
  }
}


export { Form as FormComponent };

export default connect()(Form);
