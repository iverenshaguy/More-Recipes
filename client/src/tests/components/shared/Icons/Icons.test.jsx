import React from 'react';
import { RenderIcon, RenderVoteIcon } from '../../../../components/shared/Icons';

describe('Icons', () => {
  describe('RenderIcon', () => {
    const props = {
      recipe: {
        id: 2,
        recipeName: 'Recipe',
        userId: 1,
        isReviewed: false,
        ingredients: ['rice', 'beans'],
        preparations: ['wash', 'cook'],
        directions: ['wash', 'cook'],
        User: {
          id: 1,
          username: 'name'
        }
      },
      type: 'review',
      reviewed: false,
      handleClick: jest.fn()
    };

    it('renders correctly', () => {
      const shallowWrapper = shallow(<RenderIcon {...props} />);

      expect(toJson(shallowWrapper)).toMatchSnapshot();
    });

    it('renders closed star when reviewed', () => {
      const mountedWrapper = mount(<RenderIcon
        {...props}
        icon="star"
        recipe={{ ...props.recipe, isReviewed: true }}
      />);

      expect(mountedWrapper.find('i.fa-star').length).toBeTruthy();
    });

    it('calls handleClick', () => {
      const mountedWrapper = mount(<RenderIcon {...props} icon="star" />);
      mountedWrapper.find('i.fa-star-o').simulate('click');

      expect(props.handleClick).toHaveBeenCalled();
      mountedWrapper.unmount();
    });

    it('renders correctly review icon when reviewed', () => {
      const shallowWrapper = shallow(<RenderIcon {...props} reviewed />);

      expect(toJson(shallowWrapper)).toMatchSnapshot();
      expect(shallowWrapper.find('FontAwesome[name="star"]').length).toBeTruthy();
    });
  });

  describe('RenderVoteIcon', () => {
    const props = {
      recipe: {
        id: 2,
        recipeName: 'Recipe',
        userId: 1,
        isReviewed: false,
        vote: null,
        ingredients: ['rice', 'beans'],
        preparations: ['wash', 'cook'],
        directions: ['wash', 'cook'],
        User: {
          id: 1,
          username: 'name'
        }
      },
      type: 'up',
      boolCheck: false,
      handleClick: jest.fn()
    };

    it('renders correctly', () => {
      const shallowWrapper = shallow(<RenderVoteIcon {...props} />);

      expect(toJson(shallowWrapper)).toMatchSnapshot();
    });

    it('renders closed thumbs-up when upvoted', () => {
      const mountedWrapper = mount(<RenderVoteIcon
        {...props}
        recipe={{ ...props.recipe, vote: true }}
        boolCheck
      />);

      expect(mountedWrapper.find('i.fa-thumbs-up').length).toBeTruthy();
    });

    it('calls handleClick', () => {
      const mountedWrapper = mount(<RenderVoteIcon {...props} />);
      mountedWrapper.find('i.fa-thumbs-o-up').simulate('click');

      expect(props.handleClick).toHaveBeenCalled();
    });
  });
});
