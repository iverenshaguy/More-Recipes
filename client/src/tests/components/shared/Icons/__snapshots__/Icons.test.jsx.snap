// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icons RenderIcon renders correctly 1`] = `
<span
  className="review"
>
  <FontAwesome
    name="undefined-o"
    onClick={[Function]}
    size="lg"
    tag="i"
  />
  <p
    className="text-capitalize d-none d-xs-none d-sm-none d-md-block"
  >
    review
  </p>
</span>
`;

exports[`Icons RenderIcon renders correctly review icon when reviewed 1`] = `
<span
  className="review"
>
  <FontAwesome
    name="star"
    size="lg"
    tag="i"
  />
  <p
    className="text-capitalize d-none d-xs-none d-sm-none d-md-block"
  >
    Reviewed
  </p>
</span>
`;

exports[`Icons RenderVoteIcon renders correctly 1`] = `
<span
  className="upvote"
>
  <FontAwesome
    name="thumbs-o-up"
    onClick={[Function]}
    size="lg"
    tag="i"
  />
  <p
    className="text-capitalize d-none d-xs-none d-sm-none d-md-block"
  >
    upvote
  </p>
</span>
`;
